src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -2162,6 +2162,288 @@ private int codePointX(Encoding enc, CodeRange codeRange, byte[] bytes, int p, i
         }
     }
 
+    @CoreMethod(names = "undump", taintFrom = 0)
+    @ImportStatic(StringGuards.class)
+    public abstract static class UndumpNode extends CoreMethodArrayArgumentsNode {
+        @Specialization(guards = "isAsciiCompatible(string)")
+        protected RubyString undumpAsciiCompatible(RubyString string,
+                @CachedLanguage RubyLanguage language,
+                @Cached AllocateHelperNode allocateHelperNode,
+                @Cached RopeNodes.MakeLeafRopeNode makeLeafRopeNode) {
+            // Taken from org.jruby.RubyString#undump
+            RopeBuilder outputBytes = undump(string.rope);
+
+            final Rope rope = makeLeafRopeNode
+                    .executeMake(
+                            outputBytes.getBytes(),
+                            outputBytes.getEncoding(),
+                            CR_UNKNOWN,
+                            outputBytes.getLength());
+
+            final RubyClass logicalClass = coreLibrary().stringClass;
+            final Shape shape = allocateHelperNode.getCachedShape(logicalClass);
+            final RubyString result = new RubyString(logicalClass, shape, false, false, rope);
+            allocateHelperNode.trace(result, this, language);
+            return result;
+        }
+
+        @Specialization(guards = "!isAsciiCompatible(string)")
+        protected RubyString undumpNonAsciiCompatible(RubyString string) {
+            throw new RaiseException(
+                    getContext(),
+                    getContext().getCoreExceptions().encodingCompatibilityError(
+                            "ASCII incompatible encoding: " + string.rope.encoding,
+                            this));
+        }
+
+        private RaiseException newRuntimeError(String message) {
+            return new RaiseException(getContext(), getContext().getCoreExceptions().runtimeError(message, this));
+        }
+
+        private RaiseException invalidFormat() {
+            return newRuntimeError(
+                    "invalid dumped string; not wrapped with '\"' nor '\"...\".force_encoding(\"...\")' form");
+        }
+
+        private static final byte[] FORCE_ENCODING_BYTES = ".force_encoding(\"".getBytes();
+
+        @TruffleBoundary
+        private RopeBuilder undump(Rope rope) {
```

Comment:
Could you move this method (and `undumpAfterBackslash` and other `private` methods) to `StringSupport`?
It seems best to keep that @TruffleBoundary logic outside the node as it's quite long.

Reply:
I can, but I feel like double checking it with you: both `undump` and `undumpAfterBackslash` actively use the context and the node itself for throwing errors. So to move them into `StringSupport` would require passing the context and the node in arguments, which seems inelegant. Alternatively we could throw a custom exception from the helpers and then catch and rewrap it in the node? Also seems hacky. Please let me know how you'd like me to proceed :)

Reply:
I think passing the `currentNode` and `context` is fine, it's done for many other `static` methods.

Reply:
OK! Both comments are fixed now.